src/core/jit/synth/synth.rs

```diff
@@ -226,21 +236,26 @@ impl<'a, Value: JsonLike<'a> + Clone + 'a> Synth<Value> {
             }
         };
 
-        result.map_err(|e| self.to_location_error(e, node))
+        result.map_err(|e| self.to_location_error(e, node, data_path))
     }
 
     fn to_location_error(
         &'a self,
         error: Error,
         node: &'a Field<Nested<Value>, Value>,
+        data_path: &DataPath,
     ) -> LocationError<Error> {
         // create path from the root to the current node in the fields tree
         let path = {
             let mut path = Vec::new();
-
+            let data_path = data_path.as_slice();
+            let mut idx = 0;
             let mut parent = self.plan.find_field(node.id.clone());
-
             while let Some(field) = parent {
+                if field.type_of.is_list() && idx < data_path.len() {
+                    path.push(PathSegment::Index(data_path[idx]));
+                    idx += 1;
```

Comment:
this code travels from children to parent fields to build the path, so it's built in backward, but the `data_path` is built in forward. So here you'll probably using the wrong indexes in case we have lists inside lists. Need a test for this